[RFC] libdrgn: add powerpc stack trace support

kamalesh-babulal opened the pull request:
Add powerpc specific register information required to retrive the
stack traces of the tasks on both live system and from the core dump.
It uses the existing DSL format to define platform registers and
helper functions to initial them. It also adds architecture specific
information to enable powerpc. Current enablement is on little-endian
only.

Signed-off-by: Kamalesh Babulal <kamalesh@linux.vnet.ibm.com>

osandov commented:
Wow, thanks! I'd like to test this out, do you have suggestions for a distro/userspace environment I could run in QEMU to try it?

kamalesh-babulal commented:
> Wow, thanks! I'd like to test this out, do you have suggestions for a distro/userspace environment I could run in QEMU to try it?

Thank you :) I have been running it with:
```
qemu-system-ppc64 - 4.1.91 (v4.2.0-rc1-22-ga5c2a23510)
Ubuntu 20.04.1(VM)
libkdumpfile (a83a6e528a77)
```

kamalesh-babulal commented:
@osandov were you able to try it on ppc64le?

kamalesh-babulal commented:
Hey @osandov, 
Any comments/suggestions on the patch?
Thanks.

osandov commented:
Hi, I was able to get a ppc64le VM working, so I will take a look this week.

kamalesh-babulal commented:
@osandov Did you get a chance to try it out on ppc64le?

osandov reviewed (commented):
So sorry about the delay. This worked in the basic tests I did. I have a few questions. Thanks so much!

kamalesh-babulal commented on libdrgn/arch_ppc64.c.in:
```diff
@@ -0,0 +1,323 @@
+	if (err)
+		goto out;
+
+	err = drgn_object_member(&pt_regs_obj, &pt_regs_obj, "regs");
```
No worries, thanks for the review.  I guess, it can be done, by passing the `task_struct->thread_struct->pt_regs` to  `pt_regs_set_initial_registers_ppc64()` in `linux_kernel_set_initial_registers_ppc64()`, also the code to adjust the stack pointer to skip the first frame (unreliable), needs to stay along with link register (return address) adjustment code based on the new frame.

kamalesh-babulal replied:
Something like https://gist.github.com/kamalesh-babulal/991f904246ff1975232ffb4aa02ccd80

osandov replied:
A follow-up question since I'm not familiar with ppc: what about the first frame is unreliable?

kamalesh-babulal replied:
AFAIK, The reason for bottom-most stack frame being unreliable for the scheduled out task, is that the frame is completely uninitialized and if the task was in do_IRQ() (interrupt), it will have residual of exception marker in the first frame but in the both cases complying to the "Power Architecture 64-Bit ELF V2 ABI", the Link Register (LR) will contain the back chain pointing to the previous frame.

kamalesh-babulal commented on libdrgn/arch_ppc64.c.in:
```diff
@@ -0,0 +1,323 @@
+#include "program.h"
+%}
+
+ppc64
```
AFAIK, `__powerpc64__` is used for compiler macros and `ppc64` are used as function name prefix or appended to function name, as hints. I have no strong opinion, if you prefer `powerpc64`, let me change it.

kamalesh-babulal commented on libdrgn/linux_kernel_helpers.c:
```diff
@@ -7,6 +7,18 @@
 #include "program.h"
 
+#ifdef __powerpc64__
+struct drgn_error *linux_helper_read_vm(struct drgn_program *prog,
```
Agree, the stub is the workaround for the error `virtual address translation is not implemented for ppc64 architecture` and progress with printing the stack.  Returning `NULL` from `read_memory_via_pgtable()` will help both vmcore/live kernel stack printing, would that be the right code patch, sorry still learning my way around code.

osandov replied:
I just pushed 3360170336b254d82520524c9c310c1a06e9501e which should fix this without the need for the stub.

kamalesh-babulal replied:
Thanks, I understand your review comment better. I was looking at the wrong place.

kamalesh-babulal commented:
**V2 Changes**
Addressed the following review comments from @osandov 
- Simpifed `linux_kernel_set_initial_registers_ppc64()` to use `set_initial_registers_from_struct_ppc64()`, instead of manually grabbing registers.
- Removed `linux_helper_read_vm()` stub (commit 3360170336b2 fixes it).
- Retaining the ppc64 name.

osandov commented:
I'm giving this another look and cross-referencing the Linux kernel ppc64 context switching code, and I'm confused. `arch/powerpc/kernel/entry_64.S`. `_switch` doesn't seem to use `task->thread.regs` at all. It looks like it allocates a `struct pt_regs` on the stack and saves only r14-r32, lr, and the control register. Am I missing something, or should we be getting the registers from the on-stack `struct pt_regs` (and only those which are saved)?

kamalesh-babulal commented:
Sorry, I was out of office for the annual vacation returned this week. Happy New year.

There are two arguments, those get passed to `_switch()`. First argument (register 3) points to current tasks `task_struct->thread_struct` and second argument (register 4) points to next tasks `task_struct->thread_struct`. That's right, it saves the link register to the stack and allocates a frame of size `SWITCH_FRAME_SIZE`, which is defined as `(STACK_FRAME_OVERHEAD + sizeof(struct pt_regs))`.

The non-volatile registers r14-r31 `(thread_struct->pt_regs->gpr)` gets saved to the newly allocated frame, along with `nip (thread_struct->pt_regs->nip)`, `condition register (thread_struct->pt_regs->ccr)`. Current stack pointer address is stored to `(current)thread_struct->ksp` address.

In https://elixir.bootlin.com/linux/v5.11-rc2/source/arch/powerpc/kernel/entry_64.S#L638, the r3 gets loaded with address of `task_struct` that's calculated from the structure member offset of `(current)thread_struct` in `task_struct`. Now, when we access the `thread_struct`, that points to the stack frame address allocated in `_switch` and `pt_regs` can be retrieved `(thread_struct->regs)`. It's expected that the other volatile registers is  saved by the caller.

osandov commented:
Happy New Year! I hope you had a good break.

From what I can tell, `task->thread.regs` is distinct from the registers saved by `_switch()`:

```pycon
>>> t = find_task(prog, 1)
>>> task_state_to_char(t)
'S'
>>> hex(t.thread.regs)
'0xc00000007e4ebe90'
>>> hex(t.thread.ksp + 112)
'0xc00000007e4eb950'
>>> t.thread.regs.value_() - (t.thread.ksp.value_() + 112)
1344
>>> sizeof(prog.type("struct pt_regs"))
368
```

`task->thread.regs` seems to contain the saved userspace registers:

```pycon
>>> hex(t.thread.regs.nip)
'0x7b6a9cbbeff8'
>>> hex(t.thread.regs.gpr[1])
'0x7fffe610b3d0'
```

```console
$ sudo grep 'r-x.*libc-\|\[stack\]' /proc/1/maps
7b6a9ca70000-7b6a9cc80000 r-xp 00000000 fc:02 529976                     /usr/lib/powerpc64le-linux-gnu/libc-2.31.so
7fffe60e0000-7fffe6110000 rw-p 00000000 00:00 0                          [stack]
```

So I think we need read the registers from the `struct pt_regs` at `task->thread.ksp + STACK_FRAME_OVERHEAD`. Something like:

```c
static struct drgn_error *
linux_kernel_set_initial_registers_ppc64(Dwfl_Thread *thread,
					 const struct drgn_object *task_obj)
{
	struct drgn_error *err;
	struct drgn_program *prog = drgn_object_program(task_obj);
	bool bswap;
	err = drgn_program_bswap(prog, &bswap);
	if (err)
		return err;

	struct drgn_object sp_obj;
	drgn_object_init(&sp_obj, prog);

	err = drgn_object_member_dereference(&sp_obj, task_obj, "thread");
	if (err)
		goto out;
	err = drgn_object_member(&sp_obj, &sp_obj, "ksp");
	if (err)
		goto out;
	uint64_t ksp;
	err = drgn_object_read_unsigned(&sp_obj, &ksp);
	if (err)
		goto out;
	static const uint64_t STACK_FRAME_OVERHEAD = 112;
	static const uint64_t SWITCH_FRAME_SIZE = STACK_FRAME_OVERHEAD + 368;
	char buf[312];
	err = drgn_program_read_memory(prog, buf, ksp + STACK_FRAME_OVERHEAD,
				       sizeof(buf), false);
	if (err)
		goto out;

#define READ_REGISTER(n) ({				\
	uint64_t reg;					\
	memcpy(&reg, (uint64_t *)buf + n, sizeof(reg));	\
	bswap ? bswap_64(reg) : reg;			\
})

	/* Set pc from struct pt_regs::nip. */
	dwfl_thread_state_register_pc(thread, READ_REGISTER(32));

	/* General purpose registers. */
	Dwarf_Word dwarf_regs[32 - 14];
	for (int i = 14; i < 32; i++)
		dwarf_regs[i - 14] = READ_REGISTER(i);
	if (!dwfl_thread_state_registers(thread, 14, ARRAY_SIZE(dwarf_regs),
					 dwarf_regs)) {
		err = drgn_error_libdwfl();
		goto out;
	}

	/* Stack pointer. */
	dwarf_regs[0] = ksp + SWITCH_FRAME_SIZE;
	if (!dwfl_thread_state_registers(thread, 1, 1, dwarf_regs)) {
		err = drgn_error_libdwfl();
		goto out;
	}

	/* Condition register fields. */
	uint32_t ccr = READ_REGISTER(38);
	for (int i = 0; i < 8; i++)
		dwarf_regs[i] = (ccr >> (4 * i)) & 0xf;
	if (!dwfl_thread_state_registers(thread, 68, 8, dwarf_regs)) {
		err = drgn_error_libdwfl();
		goto out;
	}

#undef READ_REGISTER

	err = NULL;
out:
	drgn_object_deinit(&sp_obj);
	return err;
}
```

Does that look reasonable?

osandov commented:
I also noticed that the register numbers defined here are inconsistent with the [spec](https://refspecs.linuxfoundation.org/ELF/ppc64/PPC-elf64abi.html#DW-REG) I found online and the numbers used in elfutils. I fixed that, cleaned up the logic a bit, rebased it, and pushed it all to https://github.com/osandov/drgn/tree/powerpc-stack-trace. Let me know if that looks alright.

kamalesh-babulal commented:
> Happy New Year! I hope you had a good break.

Thank you. It was a quite one. 

> From what I can tell, task->thread.regs is distinct from the registers saved by _switch():

I get it now, yeah your right. Re-reading along with your values printed from `_switch`, I had overlooked the `STACK_FRAME_OVERHEAD` offset, for retrieving the `pt_regs` values.

> I also noticed that the register numbers defined here are inconsistent with the spec I found online and the numbers used in elfutils. I fixed that, cleaned up the logic a bit, rebased it, and

Thanks for working on it. That's right, the `section 2.4 DWARF definition` and the values used in elfutils don't match. I had used the DWARF values recommended in `Power Architecture 64-Bit ELF V2 ABI Specification`.

>  pushed it all to https://github.com/osandov/drgn/tree/powerpc-stack-trace. Let me know if that looks alright.

Looks good to me, I ran a quick test and everything looks alright. A minor nit-pick, 
```
--- a/libdrgn/arch_ppc64.c.in
+++ b/libdrgn/arch_ppc64.c.in
@@ -125,7 +125,7 @@ set_initial_registers_from_struct_ppc64(Dwfl_Thread *thread, const void *regs,
                                                 dwarf_regs))
                        return drgn_error_libdwfl();
                dwarf_regs[0] = READ_REGISTER(36); /* link */
-               dwarf_regs[0] = READ_REGISTER(35); /* ctr */
+               dwarf_regs[1] = READ_REGISTER(35); /* ctr */
                if (!dwfl_thread_state_registers(thread, DRGN_REGISTER_PPC64_lr,
                                                 2, dwarf_regs))
                        return drgn_error_libdwfl();
```
The index for `ctr register` should be `1` instead to `0`.

osandov commented:
Huh, this is confusing. There seem to be two register number definitions floating around. The one used by elfutils seems to be older, [64-bit PowerPC ELF Application Binary Interface Supplement 1.9
](https://refspecs.linuxfoundation.org/ELF/ppc64/PPC-elf64abi.html#DW-REG). GCC and binutils use their own that seems to have been codified in [Power Architecture 64-Bit ELF V2 ABI Specification](http://cdn.openpowerfoundation.org/wp-content/uploads/resources/leabi/content/dbdoclet.50655240_82998.html). The latter is the one you used.

We need to use the one that GCC generates. I changed this back to the numbering from your PR, but limited it only to the GPRs, FPRs, lr, and cr0-cr7 to reduce confusion with the other registers that don't match. As far as I can tell, the elfutils unwinder still works even though it uses the older scheme. I pushed the updated version. Please take another look, and I'll go ahead and merge this if everything looks good.

kamalesh-babulal commented:
> Huh, this is confusing. There seem to be two register number definitions floating around. The one used by elfutils seems to be older, 64-bit PowerPC ELF Application Binary Interface Supplement 1.9
. GCC and binutils use their own that seems to have been codified in Power Architecture 64-Bit ELF V2 ABI Specification. The latter is the one you used.

There has been the inconsistency with the registers number used and elfutils `ppc_dwarf_to_regno()` fixes the register numbers.

> We need to use the one that GCC generates. I changed this back to the numbering from your PR, but limited it only to the GPRs, FPRs, lr, and cr0-cr7 to reduce confusion with the other registers that don't match. As far as I can tell, the elfutils unwinder still works even though it uses the older scheme. I pushed the updated version. Please take another look, and I'll go ahead and merge this if everything looks good.

Looks good, the stacktrace is not right just with one case, which I previously missed while testing over the weekend. Sorry my bad.  The usecase is while using the dumpfile:
```
>>> t = find_task(prog, 115205)
>>> task_state_to_char(t)
'R'
```
`115205 (bash)`, from where the `echo c > /proc/sysrq-trigger` was executed. It's the panic task.

The call path is:
```
pt_regs_set_initial_registers_ppc64()
 - set_initial_registers_from_struct_ppc64()
 ....
   dwfl_thread_state_register_pc(thread, READ_REGISTER(32)); /* nip */ 
....
>>> prog.stack_trace(115205)
#0  __crash_kexec+0xf8/0x10c
#1  0xfffffffffffffffb
#2  0xfffffffffffffffb
#3  0x2
#4  vfs_write+0xf0/0x348
#5  system_call_exception+0xf4/0x1f8
#6  system_call_common+0xf0/0x0

 ```

skipping the setting of `pc` for such tasks, the elfutils unwinds the stack trace:
```
--- a/libdrgn/arch_ppc64.c.in
+++ b/libdrgn/arch_ppc64.c.in
@@ -115,7 +115,8 @@ set_initial_registers_from_struct_ppc64(Dwfl_Thread *thread, const void *regs,
        bswap ? bswap_64(reg) : reg;                            \
 })
 
-       dwfl_thread_state_register_pc(thread, READ_REGISTER(32)); /* nip */
+       if (!complete)
+               dwfl_thread_state_register_pc(thread, READ_REGISTER(32)); /* nip */
 
        int min_gpr = complete ? 0 : 14;
        for (int i = min_gpr; i < 32; i++)

>>> prog.stack_trace(115205)
#0  panic+0x194/0x408
#1  sysrq_handle_crash+0x30/0x34
#2  __handle_sysrq+0xd0/0x23c
#3  write_sysrq_trigger+0xb4/0x10c
#4  proc_reg_write+0x100/0x138
#5  vfs_write+0xf0/0x348
#6  ksys_write+0x84/0x13c
#7  system_call_exception+0xf4/0x1f8
#8  system_call_common+0xf0/0x0
```
I need to dig more into elftuils code, to understand what is going under the hood.

kamalesh-babulal commented:
The issue with the task unwinding of running a.k.a. non-switched task is that, the right `Link Register(LR)` value is not set as the `NIP/PC`. The task is running, hence the `LR` register point to the previous frame as required by the `ELF v2 ABI specification`. The reason skipping to set PC register, unwinds the stack successfully, is because of elfutils calling `state_fetch_pc()` to set the `PC`.  `state_fetch_pc()` internally fetches the value of `reg[return address register] `which is `65` a.k.a. `link register`.

For the switched out task, the register values passed through, `ksp + SWITCH_FRAME_SIZE` sets the right `NIP(LR)` value. We can either skip the setting of `PC` or set it to register based on complete:

```
--- a/libdrgn/arch_ppc64.c.in
+++ b/libdrgn/arch_ppc64.c.in
@@ -115,7 +115,8 @@ set_initial_registers_from_struct_ppc64(Dwfl_Thread *thread, const void *regs,
        bswap ? bswap_64(reg) : reg;                            \
 })
 
-       dwfl_thread_state_register_pc(thread, READ_REGISTER(32)); /* nip */
+       int regno = (complete ? 36 : 32);
+       dwfl_thread_state_register_pc(thread, READ_REGISTER(regno)); /* nip */
 
        int min_gpr = complete ? 0 : 14;
        for (int i = min_gpr; i < 32; i++)
```

kamalesh-babulal commented:
I can push the patch with your preference on handling the running task. I have not a strong opinion on them.

osandov commented:
The kernel seems to populate `NT_PRSTATUS` oddly. See [arch/powerpc/kernel/ppc_save_regs.S](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/powerpc/kernel/ppc_save_regs.S?id=abc3fce76adbdfa8f87272c784b388cd20b46049). It sets nip to the caller's program counter but r1 to the caller's caller's stack pointer.

I think we need something like this:
```diff
diff --git a/libdrgn/arch_ppc64.c.in b/libdrgn/arch_ppc64.c.in
index 6271c3e..2398dbd 100644
--- a/libdrgn/arch_ppc64.c.in
+++ b/libdrgn/arch_ppc64.c.in
@@ -82,12 +82,15 @@ f31
 # There are two numbering schemes for the remaining registers. The original
 # definition appears to be "64-bit PowerPC ELF Application Binary Interface
 # Supplement" [1]. The GNU toolchain instead uses its own that was later
-# codified in "Power Architecture 64-Bit ELF V2 ABI Specification" [2]. We only
-# support the latter.
+# codified in "Power Architecture 64-Bit ELF V2 ABI Specification" [2]. We use
+# the latter.
+#
+# What the ABI calls the link register (register 65) is used as the DWARF CFI
+# return_address_register column in practice, which is actually the program
+# counter. We omit it to avoid confusion.
 #
 # 1: https://refspecs.linuxfoundation.org/ELF/ppc64/PPC-elf64abi.html
 # 2: https://openpowerfoundation.org/?resource_lib=64-bit-elf-v2-abi-specification-power-architecture
-lr, 65
 cr0, 68
 cr1
 cr2
@@ -100,7 +103,9 @@ cr7
 
 static struct drgn_error *
 set_initial_registers_from_struct_ppc64(Dwfl_Thread *thread, const void *regs,
-					size_t size, bool bswap, bool complete)
+					size_t size, bool bswap,
+					bool linux_kernel_prstatus,
+					bool linux_kernel_switched_out)
 {
 	if (size < 312) {
 		return drgn_error_create(DRGN_ERROR_INVALID_ARGUMENT,
@@ -111,13 +116,31 @@ set_initial_registers_from_struct_ppc64(Dwfl_Thread *thread, const void *regs,
 
 #define READ_REGISTER(n) ({					\
 	uint64_t reg;						\
-	memcpy(&reg, (uint64_t *)regs + n, sizeof(reg));	\
+	memcpy(&reg, (uint64_t *)regs + (n), sizeof(reg));	\
 	bswap ? bswap_64(reg) : reg;				\
 })
 
-	dwfl_thread_state_register_pc(thread, READ_REGISTER(32)); /* nip */
+	/*
+	 * The NT_PRSTATUS note in Linux kernel vmcores is odd. Since Linux
+	 * kernel commit d16a58f8854b ("powerpc: Improve ppc_save_regs()") (in
+	 * v5.7), the saved stack pointer (r1) is for the caller of the program
+	 * counter saved in nip (word 36). Before that, the saved nip is set to
+	 * the same as the link register (word 32). So, use the link register
+	 * instead of nip.
+	 */
+	dwarf_regs[0] = READ_REGISTER(linux_kernel_prstatus ? 36 : 32);
+	/*
+	 * The ABI calls register 65 the link register, but it's actually used
+	 * as the program counter.
+	 */
+	if (!dwfl_thread_state_registers(thread, 65, 1, dwarf_regs))
+		return drgn_error_libdwfl();
 
-	int min_gpr = complete ? 0 : 14;
+	/*
+	 * Switched out tasks in the Linux kernel only save the callee-saved
+	 * general purpose registers (14-31).
+	 */
+	int min_gpr = linux_kernel_switched_out ? 14 : 0;
 	for (int i = min_gpr; i < 32; i++)
 		dwarf_regs[i] = READ_REGISTER(i);
 	if (!dwfl_thread_state_registers(thread, min_gpr, 32 - min_gpr,
@@ -131,13 +154,6 @@ set_initial_registers_from_struct_ppc64(Dwfl_Thread *thread, const void *regs,
 					 dwarf_regs))
 		return drgn_error_libdwfl();
 
-	if (complete) {
-		dwarf_regs[0] = READ_REGISTER(36); /* link */
-		if (!dwfl_thread_state_registers(thread, DRGN_REGISTER_PPC64_lr,
-						 1, dwarf_regs))
-			return drgn_error_libdwfl();
-	}
-
 #undef READ_REGISTER
 
 	return NULL;
@@ -152,7 +168,7 @@ pt_regs_set_initial_registers_ppc64(Dwfl_Thread *thread,
 	return set_initial_registers_from_struct_ppc64(thread,
 						       drgn_object_buffer(obj),
 						       drgn_object_size(obj),
-						       bswap, true);
+						       bswap, false, false);
 }
 
 static struct drgn_error *
@@ -168,9 +184,11 @@ prstatus_set_initial_registers_ppc64(struct drgn_program *prog,
 	struct drgn_error *err = drgn_program_bswap(prog, &bswap);
 	if (err)
 		return err;
+	bool is_linux_kernel = prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL;
 	return set_initial_registers_from_struct_ppc64(thread,
 						       (char *)prstatus + 112,
-						       size - 112, bswap, true);
+						       size - 112, bswap,
+						       is_linux_kernel, false);
 }
 
 static struct drgn_error *
@@ -209,7 +227,7 @@ linux_kernel_set_initial_registers_ppc64(Dwfl_Thread *thread,
 
 	err = set_initial_registers_from_struct_ppc64(thread, regs,
 						      sizeof(regs), bswap,
-						      false);
+						      false, true);
 	if (err)
 		goto out;
 
```

kamalesh-babulal commented:
> The kernel seems to populate NT_PRSTATUS oddly. See arch/powerpc/kernel/ppc_save_regs.S. It sets nip to the caller's program counter but r1 to the caller's caller's stack pointer.
>I think we need something like this:

`ppc_save_regs` explains why `LR` should be used instead of `NIP` for the tasks initializing their registers values via `pt_regs_set_initial_registers_ppc64()`.  Do you want me to force to push the PR along with this delta? (The above code diff is pasted thrice).

osandov commented:
I applied the fix and merged, thanks so much for your patience!

kamalesh-babulal commented:
> I applied the fix and merged, thanks so much for your patience!

@osandov Thank you so much for all the help and improvements to the code.
